### PR TITLE
Issue 3234: use model schema type instead of class definition schema

### PR DESCRIPTION
**Pull request by frank-dong-ms**
address issue: https://github.com/dotnet/machinelearning/issues/3234

### CODE REVIEW

File: test/Microsoft.ML.Functional.Tests/ModelFiles.cs [hunk 1]
```diff
@@ -217,6 +217,7 @@ public class ModelInput
             public string[] CategoricalFeatures;
             public float[] NumericalFeatures;
 #pragma warning restore SA1401
+            public float Label;
```
Comment (frank-dong-ms) [inline test/Microsoft.ML.Functional.Tests/ModelFiles.cs:220]: >Label [](start = 25, length = 5)

add extra column Label as the model modelwithoptionalcolumntransform's schema contains an extra Label column

File: src/Microsoft.ML.Data/DataView/DataViewConstructionUtils.cs [hunk 2]
```diff
@@ -71,6 +71,13 @@ internal static SchemaDefinition GetSchemaDefinition<TRow>(IHostEnvironment env,
                 var schemaDefinitionCol = schemaDefinition.FirstOrDefault(c => c.ColumnName == name);
                 if (schemaDefinitionCol == null)
                     throw env.Except($"Type should contain a member named {name}");
+
+                //Always use column type from model as this type can be more specific.
+                //This can be corner case:
+                //For example, we can load an model whose schema contains Vector<Single, 38>
```
Comment (mstfbl) [inline src/Microsoft.ML.Data/DataView/DataViewConstructionUtils.cs:77]: Will there be a check or at least logging for this corner case? Maybe we should put in a check for arrays of size 0, which is what `float[]` would be. Also is there a check for the size of provided arrays matching the size of the corresponding `Vector`?

Comment (frank-dong-ms) [reply src/Microsoft.ML.Data/DataView/DataViewConstructionUtils.cs:77]: test LoadModelWithOptionalColumnTransform is to test this case. 
for length check, it is applied at below:
[https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Scorers/SchemaBindablePredictorWrapper.cs#L129-L147](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Scorers/SchemaBindablePredictorWrapper.cs#L129-L147)

---
In reply to: [439585282](https://github.com/dotnet/machinelearning/pull/5228#discussion_r439585282) [](ancestors = 439585282)

File: test/Microsoft.ML.Functional.Tests/ModelFiles.cs [hunk 3]
```diff
@@ -230,15 +231,17 @@ public class ModelOutput
         [Fact]
         public void LoadModelWithOptionalColumnTransform()
         {
```
Comment (harishsk) [inline test/Microsoft.ML.Functional.Tests/ModelFiles.cs:233]: Does this fix invalidate the old test?

It appears that the the previous test was testing this using SchemaDefinition and to reproduce this issue you are using the one DataViewSchema. 

Does it make sense to test both code paths not replace the previous one?

Comment (frank-dong-ms) [reply test/Microsoft.ML.Functional.Tests/ModelFiles.cs:233]: No, the old test is still valid. I'm ok with test both code path not just replace as I don't find a test is test the previous case directly.

This test was originally created at below PR:
[https://github.com/dotnet/machinelearning/pull/3700](https://github.com/dotnet/machinelearning/pull/3700)

From the conversation you can see the original intention was to use the modified way to write test(load model with schema output then use the schema to create prediction engine), but with the type bug this test was written in the way with extra schema definition.

I will made the change to test both ways of create prediction engine, thanks


---
In reply to: [440394844](https://github.com/dotnet/machinelearning/pull/5228#discussion_r440394844) [](ancestors = 440394844)

**Pull request merged**